### ISSUE TITLE
Nullability of return type of anonymous function in overrides ignored

### ISSUE THREAD

**Issue by kevinoconnor7** [google/closure-compiler#2009]
I ran into a case where someone was overriding a method that takes an anonymous function as a param and weakened its return type to be nullable. Curious as to how this happened I quickly discovered that the compiler doesn't emit an error for this.

For example:

``` javascript
class A {
  /** @param {function() : !Object} bar */
  foo(bar) {}
}

class B extends A {
  /**
   * @override
   * @param {function() : ?Object} bar
   */
  foo(bar) {}
}
```

Here is the clean result:
https://closure-compiler-debugger.appspot.com/#input0%3Dclass%2520A%2520%257B%250A%2520%2520%252F**%2520%2540param%2520%257Bfunction()%2520%253A%2520!Object%257D%2520bar%2520*%252F%250A%2520%2520foo(bar)%2520%257B%257D%250A%257D%250A%250Aclass%2520B%2520extends%2520A%2520%257B%250A%2520%2520%252F**%250A%2520%2520%2520*%2520%2540override%250A%2520%2520%2520*%2520%2540param%2520%257Bfunction()%2520%253A%2520%253FObject%257D%2520bar%250A%2520%2520%2520*%252F%250A%2520%2520foo(bar)%2520%257B%257D%250A%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_REPORT_MISSING_OVERRIDE%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

NTI also does not catch this case.

**Comment by dimvar** [comment]
This is working as intended. B#foo must be a subtype of A#foo, which means that the arguments of B#foo must be supertypes of the arguments of A#foo. And the function that returns `?Object` is a supertype of the function that returns `!Object`.

**Comment by kevinoconnor7** [comment]
Woops, I simplified a bit too much when writing it in ES6. The case I actually encountered was more along the lines of:

```javascript
/** @constructor */
function A() {};

/** @param {function() : !Object} bar */
A.prototype.foo = function(bar) {};


/** @constructor @extends {A} */
function B() {};

/**
 * @override
 * @param {function() : ?Object} bar
 */
B.prototype.foo; // Equivalent to just calling the super.

/** @return {?Object} */
function Bar() { return null; };

new B().foo(Bar);
```

**Comment by dimvar** [comment]
What is the difference in the new code? It is valid to declare `B#foo` with the new type.

**Comment by Dominator008** [comment]
@dimvar I think @kevinoconnor7 meant that by calling `new B().foo(Bar);`, we are actually invoking `A.prototype.foo` which takes a non-nullable `Object`.

**Comment by kevinoconnor7** [comment]
@Dominator008 is correct. The second example makes it such that `B.foo` implicitly calls the super.

In the ES6 example, if `B.foo` called `super.foo(bar)`, that would result in a type error. In the second example I have essentially done that but got around the error.

**Comment by dimvar** [comment]
I see. So, the bug is that we shouldn't allow narrowing the type of a method if the method definition is a stub. That seems reasonable to me. I'll reopen and fix in NTI. No promises about the old type checker.

**Comment by kevinoconnor7** [comment]
@concavelenz Was this actually fixed?

**Comment by concavelenz** [comment]
I bulk closed all the NTI issues.  Though, here this is working as intended.  This is just how function subtyping works.   We had a nice post about it years ago:

http://closuretools.blogspot.com/2012/06/subtyping-functions-without-poking-your.html
